### ISSUE TITLE
`unnecessary_lazy_evaluations`: lint `{Option,Result}::map_or_else`

### ISSUE THREAD

**Issue by ada4a** [rust-lang/rust-clippy#15829]
### Description

None of the following currently get linted:
(EDIT: this used to include `map_or`, but as the comments below rightfully say, its closure can't actually be made eager)
```rs
#![allow(unused)]
#![warn(clippy::unnecessary_lazy_evaluations)]

fn main() {
    let o = Some(42);
    o.map_or_else(|| 42, |_| 42);
    
    let r: Result<i32, i32> = Ok(42);
    r.map_or_else(|_| 42, |_| 42);
}
```

It would be cool if they did:
```rs
#![allow(unused)]
#![warn(clippy::unnecessary_lazy_evaluations)]

fn main() {
    let o = Some(42);
    o.map_or(42, |_| 42);
    
    let r: Result<i32, i32> = Ok(42);
    r.map_or(42, |_| 42);
}
```
### Version

```text
rustc 1.92.0-nightly (839222065 2025-10-05)
binary: rustc
commit-hash: 839222065a44ac21c15df68ed2f2f3c0127b0b8e
commit-date: 2025-10-05
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.2
```

### Additional Labels

@rustbot label C-enhancement "good first issue"

**Comment by teofr** [comment]
A quick question here on what's expected of this issue.

From `o.map_or_else(|| a, |_| b);` I'd expect `unnecessary_lazy_evaluations` to suggest `o.map_or(a, |_| b)`, regardless of the values of `a` and `b`.

Is the expectation that `o.map_or(42, |_| 42);` is suggested to become just `42`?

**Comment by vangata-ve** [comment]
> From o.map_or_else(|| a, |_| b); I'd expect unnecessary_lazy_evaluations to suggest o.map_or(a, |_| b), regardless of the values of a and b.

The lint exists to catch usage of lazy evaluated closures when theres no actual point from laziness. In your example, the outcome should from `o.map_or_else(|| a, |_| b)` should be `o.map_or_else(a, |_| b)`. 

> Is the expectation that o.map_or(42, |_| 42); is suggested to become just 42?

No, I don't think it should, though don't take my word for it

EDIT: Fix a typo - `o.map_or(a, |_| b)` to `o.map_or_else(a, |_| b)`

**Comment by ada4a** [comment]
Ah, you're both right -- `map_or` shouldn't get linted at all. I got carried away by all the closures and therefore thought the "map" closure (`|_| 42`) can get replaced with `42` as well, but that's obviously not the case.

The `map_or_else` part still stands though -- as, again, both of you said

**Comment by apekros** [comment]
@vangata-ve 
Are you still planning on doing this? Otherwise I'd love to claim it :)

**Comment by vangata-ve** [comment]
> **Error**: Cannot release another user's assignment

Oh wait I'm not assigned anymore. Good

**Comment by sanskarRaghuvanshi** [comment]
Hi @ada4a! Is this issue still available? I'd like to work on it.

**Comment by ada4a** [comment]
Hi @sanskarRaghuvanshi, thank you for your interest:) There has been a PR by @apekros, but it wasn't touched in some time, so let me ask them:

Hi @apekros, do you still intend to work on that PR? If you do, just let me know – otherwise, I'd propose @sanskarRaghuvanshi takes over it (by creating a copy of the PR and addressing the review)

**Comment by ada4a** [comment]
Hm, no response... feel free to take over @sanskarRaghuvanshi

**Comment by apekros** [comment]
> Hm, no response... feel free to take over [@sanskarRaghuvanshi](https://github.com/sanskarRaghuvanshi)

Apologies for no response, feel free to take over as I am currently super busy, will pick it up again if it's still free later

**Comment by ada4a** [comment]
No worries @apekros, thank you for responding:)
